File: backend/src/main/java/com/devlaunch/service/interfaces/JobApplicationService.java
package com.devlaunch.service.interfaces;

import com.devlaunch.dto.request.CreateInterviewNoteRequest;
import com.devlaunch.dto.request.CreateJobApplicationRequest;
import com.devlaunch.dto.request.ScheduleInterviewRequest;
import com.devlaunch.dto.request.UpdateInterviewScheduleRequest;
import com.devlaunch.dto.request.UpdateJobApplicationRequest;
import com.devlaunch.dto.response.ApplicationAnalyticsResponse;
import com.devlaunch.dto.response.ApplicationAttachmentResponse;
import com.devlaunch.dto.response.InterviewNoteResponse;
import com.devlaunch.dto.response.InterviewScheduleResponse;
import com.devlaunch.dto.response.JobApplicationResponse;
import com.devlaunch.dto.response.TimelineEventResponse;
import com.devlaunch.entity.enums.ApplicationStatus;
import com.devlaunch.entity.enums.AttachmentCategory;
import com.devlaunch.exception.ResourceNotFoundException;
import org.springframework.core.io.Resource;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

/**
 * Service interface for job application tracking operations.
 * <p>
 * Defines the contract for creating, retrieving, updating, and deleting
 * the currently authenticated user's job applications, plus the placement
 * management features layered on top: status moves (Kanban), the automatic
 * application timeline, interview scheduling, private interview notes,
 * document attachments, and an analytics overview. Every operation is
 * scoped to the authenticated user, and update/delete operations verify
 * ownership before touching any record.
 * </p>
 *
 * @author DevLaunch
 */
public interface JobApplicationService {

    /**
     * Creates a new job application for the currently authenticated user.
     * <p>
     * If a resume ID is provided in the request, the resume is fetched
     * and verified to belong to the authenticated user before being
     * associated with the application. The initial timeline entries are
     * recorded automatically.
     * </p>
     *
     * @param request the create-job-application request containing application details
     * @return the newly created job application data
     * @throws ResourceNotFoundException if the authenticated user is not found,
     *                                   or if the specified resume is not found
     *                                   or does not belong to the user
     */
    JobApplicationResponse createJobApplication(CreateJobApplicationRequest request);

    /**
     * Retrieves all job applications belonging to the currently authenticated user.
     *
     * @return a list of job application data for the authenticated user,
     *         or an empty list if none exist
     */
    List<JobApplicationResponse> getAllJobApplications();

    /**
     * Retrieves a specific job application by its ID, ensuring it belongs to the
     * currently authenticated user.
     *
     * @param id the job application ID
     * @return the job application data
     * @throws ResourceNotFoundException if the job application is not found
     *                                    or does not belong to the user
     */
    JobApplicationResponse getJobApplicationById(Long id);

    /**
     * Updates a specific job application by its ID, ensuring it belongs to the
     * currently authenticated user.
     * <p>
     * If the resume ID in the request differs from the current value, the
     * resume association is updated accordingly. When the status changes,
     * a timeline event is recorded and the user is notified.
     * </p>
     *
     * @param id      the job application ID to update
     * @param request the update request containing the new application details
     * @return the updated job application data
     * @throws ResourceNotFoundException if the job application is not found
     *                                    or does not belong to the user, or if
     *                                    the specified resume is not found or
     *                                    does not belong to the user
     */
    JobApplicationResponse updateJobApplication(Long id, UpdateJobApplicationRequest request);

    /**
     * Updates only the status of a job application (used by the Kanban board
     * drag-and-drop), recording a timeline event and notification.
     *
     * @param id     the job application ID
     * @param status the new status
     * @return the updated job application data
     * @throws ResourceNotFoundException if the job application is not found
     *                                    or does not belong to the user
     */
    JobApplicationResponse updateApplicationStatus(Long id, ApplicationStatus status);

    /**
     * Deletes a specific job application by its ID, ensuring it belongs to the
     * currently authenticated user.
     *
     * @param id the job application ID to delete
     * @throws ResourceNotFoundException if the job application is not found
     *                                    or does not belong to the user
     */
    void deleteJobApplication(Long id);

    /**
     * Retrieves the milestone timeline of a job application, oldest first.
     *
     * @param id the job application ID
     * @return the ordered timeline events
     */
    List<TimelineEventResponse> getTimeline(Long id);

    /**
     * Retrieves the scheduled interviews of a job application, ordered by date.
     *
     * @param id the job application ID
     * @return the interviews, or an empty list if none exist
     */
    List<InterviewScheduleResponse> getInterviews(Long id);

    /**
     * Schedules a new interview on a job application, recording a timeline
     * event and notifying the user.
     *
     * @param id      the job application ID
     * @param request the interview details
     * @return the created interview
     */
    InterviewScheduleResponse scheduleInterview(Long id, ScheduleInterviewRequest request);

    /**
     * Updates an existing interview schedule, verifying the interview belongs
     * to the authenticated user's application.
     *
     * @param interviewId the interview ID
     * @param request     the updated interview details
     * @return the updated interview
     */
    InterviewScheduleResponse updateInterview(Long interviewId, UpdateInterviewScheduleRequest request);

    /**
     * Cancels an existing interview schedule, recording a timeline event and
     * notifying the user.
     *
     * @param interviewId the interview ID
     * @return the cancelled interview
     */
    InterviewScheduleResponse cancelInterview(Long interviewId);

    /**
     * Retrieves the private interview notes of a job application, newest first.
     *
     * @param id the job application ID
     * @return the notes, or an empty list if none exist
     */
    List<InterviewNoteResponse> getNotes(Long id);

    /**
     * Adds a private interview note entry to a job application.
     *
     * @param id      the job application ID
     * @param request the note content
     * @return the created note
     */
    InterviewNoteResponse addNote(Long id, CreateInterviewNoteRequest request);

    /**
     * Deletes a private interview note from a job application.
     *
     * @param applicationId the job application ID
     * @param noteId        the note ID to delete
     */
    void deleteNote(Long applicationId, Long noteId);

    /**
     * Retrieves the attachments of a job application, newest first.
     *
     * @param id the job application ID
     * @return the attachments, or an empty list if none exist
     */
    List<ApplicationAttachmentResponse> getAttachments(Long id);

    /**
     * Uploads and attaches a document to a job application, recording a
     * timeline event.
     *
     * @param id       the job application ID
     * @param file     the uploaded file
     * @param category the attachment category
     * @return the created attachment metadata
     */
    ApplicationAttachmentResponse uploadAttachment(Long id, MultipartFile file,
                                                   AttachmentCategory category);

    /**
     * Deletes an attachment, removing both the database record and the
     * underlying file from storage.
     *
     * @param applicationId the job application ID
     * @param attachmentId  the attachment ID to delete
     */
    void deleteAttachment(Long applicationId, Long attachmentId);

    /**
     * Prepares an attachment for download.
     *
     * @param applicationId the job application ID
     * @param attachmentId  the attachment ID to download
     * @return the downloadable file descriptor (resource, original name, content type)
     */
    AttachmentDownload downloadAttachment(Long applicationId, Long attachmentId);

    /**
     * Computes the analytics overview for the authenticated user's applications.
     *
     * @return the aggregated analytics
     */
    ApplicationAnalyticsResponse getAnalytics();

    /**
     * Immutable descriptor of a downloadable attachment file.
     *
     * @param resource    the file content as a Spring resource
     * @param fileName    the original file name for the download header
     * @param contentType the MIME content type
     */
    record AttachmentDownload(Resource resource, String fileName, String contentType) {
    }

}

File: frontend/src/services/job-application.service.ts
/**
 * Job Application service — CRUD operations for job application tracking,
 * plus the placement management features: quick status moves (Kanban),
 * timeline, interview scheduling, private notes, attachments, and analytics.
 *
 * Communicates with the backend JobApplicationController.
 *
 * @see backend/src/main/java/com/devlaunch/controller/JobApplicationController.java
 * @author DevLaunch
 */

import apiClient from '../api/client';
import { JOB_APPLICATIONS } from '../api/endpoints';
import type {
  JobApplicationResponse,
  CreateJobApplicationRequest,
  UpdateJobApplicationRequest,
  UpdateApplicationStatusRequest,
  ScheduleInterviewRequest,
  UpdateInterviewScheduleRequest,
  ApplicationStatusEnum,
  TimelineEvent,
  InterviewSchedule,
  InterviewNote,
  ApplicationAttachment,
  AttachmentCategory,
  ApplicationAnalytics,
} from '../types/job-application';

export const jobApplicationService = {
  /** POST /api/job-applications — Create a new job application. */
  create: (data: CreateJobApplicationRequest) =>
    apiClient.post<JobApplicationResponse>(JOB_APPLICATIONS.BASE, data).then((r) => r.data),

  /** GET /api/job-applications — Get all job applications for the authenticated user. */
  getAll: () =>
    apiClient.get<JobApplicationResponse[]>(JOB_APPLICATIONS.BASE).then((r) => r.data),

  /** GET /api/job-applications/{id} — Get a single job application by ID. */
  getById: (id: number) =>
    apiClient.get<JobApplicationResponse>(JOB_APPLICATIONS.BY_ID(id)).then((r) => r.data),

  /** PUT /api/job-applications/{id} — Update a job application. */
  update: (id: number, data: UpdateJobApplicationRequest) =>
    apiClient.put<JobApplicationResponse>(JOB_APPLICATIONS.BY_ID(id), data).then((r) => r.data),

  /** PUT /api/job-applications/{id}/status — Update only the status (Kanban drag). */
  updateStatus: (id: number, status: ApplicationStatusEnum) =>
    apiClient
      .put<JobApplicationResponse>(JOB_APPLICATIONS.STATUS(id), {
        status,
      } satisfies UpdateApplicationStatusRequest)
      .then((r) => r.data),

  /** DELETE /api/job-applications/{id} — Delete a job application. */
  delete: (id: number) =>
    apiClient.delete<string>(JOB_APPLICATIONS.BY_ID(id)).then((r) => r.data),

  /** GET /api/job-applications/analytics — Aggregated application analytics. */
  getAnalytics: () =>
    apiClient.get<ApplicationAnalytics>(JOB_APPLICATIONS.ANALYTICS).then((r) => r.data),

  /** GET /api/job-applications/{id}/timeline — Application milestone timeline. */
  getTimeline: (id: number) =>
    apiClient.get<TimelineEvent[]>(JOB_APPLICATIONS.TIMELINE(id)).then((r) => r.data),

  /** GET /api/job-applications/{id}/interviews — Interview schedules. */
  getInterviews: (id: number) =>
    apiClient.get<InterviewSchedule[]>(JOB_APPLICATIONS.INTERVIEWS(id)).then((r) => r.data),

  /** POST /api/job-applications/{id}/interviews — Schedule an interview. */
  scheduleInterview: (id: number, data: ScheduleInterviewRequest) =>
    apiClient
      .post<InterviewSchedule>(JOB_APPLICATIONS.INTERVIEWS(id), data)
      .then((r) => r.data),

  /** PUT /api/job-applications/interviews/{interviewId} — Update an interview. */
  updateInterview: (interviewId: number, data: UpdateInterviewScheduleRequest) =>
    apiClient
      .put<InterviewSchedule>(JOB_APPLICATIONS.INTERVIEW_BY_ID(interviewId), data)
      .then((r) => r.data),

  /** DELETE /api/job-applications/interviews/{interviewId} — Cancel an interview. */
  cancelInterview: (interviewId: number) =>
    apiClient
      .delete<InterviewSchedule>(JOB_APPLICATIONS.INTERVIEW_BY_ID(interviewId))
      .then((r) => r.data),

  /** GET /api/job-applications/{id}/notes — Private interview notes. */
  getNotes: (id: number) =>
    apiClient.get<InterviewNote[]>(JOB_APPLICATIONS.NOTES(id)).then((r) => r.data),

  /** POST /api/job-applications/{id}/notes — Add a private interview note. */
  addNote: (id: number, content: string) =>
    apiClient
      .post<InterviewNote>(JOB_APPLICATIONS.NOTES(id), { content })
      .then((r) => r.data),

  /** DELETE /api/job-applications/{id}/notes/{noteId} — Delete a note. */
  deleteNote: (id: number, noteId: number) =>
    apiClient.delete<string>(JOB_APPLICATIONS.NOTE_BY_ID(id, noteId)).then((r) => r.data),

  /** GET /api/job-applications/{id}/attachments — Attached documents. */
  getAttachments: (id: number) =>
    apiClient
      .get<ApplicationAttachment[]>(JOB_APPLICATIONS.ATTACHMENTS(id))
      .then((r) => r.data),

  /** POST /api/job-applications/{id}/attachments — Upload an attachment. */
  uploadAttachment: (id: number, file: File, category: AttachmentCategory) => {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('category', category);
    return apiClient
      .post<ApplicationAttachment>(JOB_APPLICATIONS.ATTACHMENTS(id), formData, {
        headers: { 'Content-Type': 'multipart/form-data' },
        timeout: 60_000,
      })
      .then((r) => r.data);
  },

  /** DELETE /api/job-applications/{id}/attachments/{attachmentId} — Delete an attachment. */
  deleteAttachment: (id: number, attachmentId: number) =>
    apiClient
      .delete<string>(JOB_APPLICATIONS.ATTACHMENT_BY_ID(id, attachmentId))
      .then((r) => r.data),

  /**
   * GET /api/job-applications/{id}/attachments/{attachmentId}/download —
   * Downloads an attachment as a blob and triggers a browser save.
   */
  downloadAttachment: async (id: number, attachmentId: number, fileName: string) => {
    const response = await apiClient.get<Blob>(
      JOB_APPLICATIONS.ATTACHMENT_DOWNLOAD(id, attachmentId),
      { responseType: 'blob' },
    );
    const blobUrl = window.URL.createObjectURL(response.data);
    const link = document.createElement('a');
    link.href = blobUrl;
    link.download = fileName;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    window.URL.revokeObjectURL(blobUrl);
  },
};
